cobc/output_tree.c

```diff
@@ -0,0 +1,2122 @@
+	case CB_CATEGORY_BOOLEAN: s = "CATEGORY_BOOLEAN"; break;
+	case CB_CATEGORY_INDEX: s = "CATEGORY_INDEX"; break;
+	case CB_CATEGORY_NATIONAL: s = "CATEGORY_NATIONAL"; break;
+	case CB_CATEGORY_NATIONAL_EDITED: s = "CATEGORY_NATIONAL_EDITED"; break;
+	case CB_CATEGORY_NUMERIC: s = "CATEGORY_NUMERIC"; break;
+	case CB_CATEGORY_NUMERIC_EDITED: s = "CATEGORY_NUMERIC_EDITED"; break;
+	case CB_CATEGORY_OBJECT_REFERENCE: s = "CATEGORY_OBJECT_REFERENCE"; break;
+	case CB_CATEGORY_DATA_POINTER: s = "CATEGORY_DATA_POINTER"; break;
+	case CB_CATEGORY_PROGRAM_POINTER: s = "CATEGORY_PROGRAM_POINTER"; break;
+	case CB_CATEGORY_FLOATING_EDITED: s = "CATEGORY_FLOATING_EDITED"; break;
+	case CB_CATEGORY_ERROR: s = "CATEGORY_ERROR"; break;
+	}
+
+	output_constr_string ("cb_category", s);
+}
+
+#if 0
+const char* cb_explain_class (enum cb_class x)
+{
+	switch (x){
+	case CB_CLASS_UNKNOWN: return "CLASS_UNKNOWN";		/* 0 */
+	case CB_CLASS_ALPHABETIC: return "CLASS_ALPHABETIC";		/* 1 */
+	case CB_CLASS_ALPHANUMERIC: return "CLASS_ALPHANUMERIC";		/* 2 */
+	case CB_CLASS_BOOLEAN: return "CLASS_BOOLEAN";		/* 3 */
+	case CB_CLASS_INDEX: return "CLASS_INDEX";			/* 4 */
+	case CB_CLASS_NATIONAL: return "CLASS_NATIONAL";		/* 5 */
+	case CB_CLASS_NUMERIC: return "CLASS_NUMERIC";		/* 6 */
+	case CB_CLASS_OBJECT: return "CLASS_OBJECT";		/* 7 */
+	case CB_CLASS_POINTER: return "CLASS_POINTER";		/* 8 */
+	}
+	return "CB_CLASS_UNKNOWN";
+}
+#endif
+
+/* Storage sections */
+static
+const char* string_of_cb_storage (enum cb_storage x)
+{
+	switch (x){
+	case CB_STORAGE_CONSTANT: return "CB_STORAGE_CONSTANT";	/* Constants */
+	case CB_STORAGE_FILE: return "STORAGE_FILE";		/* FILE SECTION */
+	case CB_STORAGE_WORKING: return "STORAGE_WORKING";		/* WORKING-STORAGE SECTION */
+	case CB_STORAGE_LOCAL: return "STORAGE_LOCAL";		/* LOCAL-STORAGE SECTION */
+	case CB_STORAGE_LINKAGE: return "STORAGE_LINKAGE";		/* LINKAGE SECTION */
+	case CB_STORAGE_SCREEN: return "STORAGE_SCREEN";		/* SCREEN SECTION */
+	case CB_STORAGE_REPORT: return "STORAGE_REPORT";		/* REPORT SECTION */
+	case CB_STORAGE_COMMUNICATION: return "STORAGE_COMMUNICATION";	/* COMMUNICATION SECTION */
+	}
+	return "CB_STORAGE_UNKNOWN";
+}
+
+static
+void output_cb_storage (enum cb_storage x)
+{
+	output_constr_string ( "cb_storage", string_of_cb_storage (x));
+}
+
+static
+const char* string_of_cob_statement (enum cob_statement x)
+{
+	switch (x){
+	case STMT_UNKNOWN: return "STMT_UNKNOWN";
+#define COB_STATEMENT(stmt, string) case stmt: return string;
+#include "../libcob/statement.def"	/* located and installed next to common.h */
+	case STMT_MAX_ENTRY: return "STMT_MAX_ENTRY";
+	}
+	return "COB_STATEMENT_UNKNOWN";
+}
+
+static
+void output_cob_statement (enum cob_statement x)
+{
+	output_constr_string ( "cob_statement", string_of_cob_statement (x));
+}
+
+
+static
+const char* string_of_cb_usage (enum cb_usage x)
+{
+	switch (x){
+	case CB_USAGE_BINARY: return "CB_USAGE_BINARY";		/* 0 */
```

Comment:
Couldn't those and similar be "stringified" using our existing macros? This would still need a switch-case, but if the source is a macro then it can be just defined as additional macro instead of the full switch-case, otherwise a macro can still be used to only end in `CASE_STRING(some-enum)`.

Note: you don't need to include the numeric values (which possibly be changed later on in any case).